### ISSUE TITLE
no way to make a dep between custom_target() and preprocess()

### ISSUE THREAD

**Issue by stsp** [mesonbuild/meson#12715]
### Discussed in https://github.com/mesonbuild/meson/discussions/12711

<div type='discussions-op-text'>

<sup>Originally posted by **stsp** January  8, 2024</sup>
I have this code:
```
plt_inc = custom_target('plt.inc',
  output: 'plt.inc',
  input: tct,
  command: [PD, '3', '@INPUT@'],
  capture: true)
cpp = meson.get_compiler('c')
sfiles = cpp.preprocess(['plt.S'],
  output: '@BASENAME@.s',
  dependencies: declare_dependency(
    compile_args: ['-D__ASSEMBLER__']))
```

Both targets work properly separately,
when started by hands in a needed order.
But I need to make sure `plt_inc` is built
before running the preprocessor for `plt.S`.
No matter what I do, this ordering is not
achieved.
How can I do that?</div>

**Comment by bruchar1** [comment]
It shouldn't be difficult to add a `depends` keyword argument to `compiler.preprocess`. You could open a pull request for that.

Btw, I think you could pass `compile_args` directly to the `preprocess` function, without the need of an internal dependency.

**Comment by stsp** [comment]
There is already some
https://mesonbuild.com/Reference-manual_returned_compiler.html#preprocess_dependencies
`dependencies` keyword.
Oh well, needs to add also `depends`. :)
I have no idea about the differences of such.

> Btw, I think you could pass compile_args directly to the preprocess function,
> without the need of an internal dependency.

And I think that meson is full of bugs. :)
So when something works, its better not
to touch.
If I do this:
```
sfiles = cpp.preprocess('plt.S',
  output: '@BASENAME@.s',
  compile_args: ['-D__ASSEMBLER__'])
```
then the result is:
```
-E -P -xc -Isrc/preprocessor_0.p -Isrc -I../src -fdiagnostics-color=always
-D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -MD -MQ src/preprocessor_0.p/plt.s
-MF src/preprocessor_0.p/plt.s.d -o src/preprocessor_0.p/plt.s -c ../src/plt.S
```
So doesn't work.

If I do this:
```
sfiles = cpp.preprocess(['plt.S'],
  output: '@BASENAME@.s',
  dependencies: declare_dependency(
    compile_args: ['-D__ASSEMBLER__']))
```
Then I get:
```
-E -P -xc -Isrc/preprocessor_0.p -Isrc -I../src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -D__ASSEMBLER__ -MD -MQ src/preprocessor_0.p/plt.s -MF src/preprocessor_0.p/plt.s.d -o src/preprocessor_0.p/plt.s -c ../src/plt.S
```

So that way it works.
Why?
You tell me. :) I have no idea.
My current strategy is when something
suddenly works, not to touch, because
for me meson is entirely unpredictable
(I am very new to it).

**Comment by stsp** [comment]
> It shouldn't be difficult to add a depends keyword argument to compiler.preprocess.
> You could open a pull request for that.

I tried this patch:
[a.diff.txt](https://github.com/mesonbuild/meson/files/13914560/a.diff.txt)
`depends` is now accepted, but doesn't
work as expected. Could you please tell
me what modification is missing here?

**Comment by bruchar1** [comment]
The patch looks good. I think the keyword added to the BuildTarget constructor should be `link_depends' instead of 'depends'.

Writing a test could help you debugging this (and will be required for your improvement being accepted). You could try adding it to 'test cases\common\259 preprocess\'

**Comment by stsp** [comment]
[0001-depends.patch.txt](https://github.com/mesonbuild/meson/files/13927952/0001-depends.patch.txt)
Here's the patch with the test-case.
It prints warnings during configure.

**Comment by stsp** [comment]
> `link_depends' instead of 'depends'.

Thanks, that gets rid of the warning.
Updated patch here.
[0001-depends.patch.txt](https://github.com/mesonbuild/meson/files/13935619/0001-depends.patch.txt)

So any help at seeing why the patch
doesn't work?

**Comment by stsp** [comment]
https://github.com/mesonbuild/meson/pull/12743
Here's the PR.
By doesn't work I mean that the
dependency target doesn't run before
our target. And the added test-case fails.

**Comment by bruchar1** [comment]
My mistake. 'link_depends' was a wrong path... There is no link involved in preprocess.

Looking further, I realize it is a little more complicated than I thought... For a build target, you would simply pass the generated header as a source. But in this particular case, passing the header as a source would result in calling the preprocessor on that header, without resolving the correct order.

**Comment by stsp** [comment]
> My mistake. 'link_depends' was a wrong path... There is no link involved in preprocess.

'depends' was also not possible there,
since it is not recognized by `BuildTarget`,
so that generated a warning.
So pass nothing to `BuildTarget` then?
Anyway, I don't think I know what to do
next.

**Comment by bruchar1** [comment]
Fixed in #12743
